rascal-lsp/src/main/rascal/util/LanguageServer.rsc

```diff
@@ -125,204 +109,132 @@ of the user.
-* ((inlayHinter)) discovers plances to add "inlays" (little views embedded in the editor on the same line). Unlike ((lenses)) inlays do not offer command execution.
-* ((executor)) executes the commands registered by ((lenses)) and ((inlayHinter))s.
+   * a ((hover)) service is a fast and location specific version of the `documentation` relation in a ((Summary)).
+   * a ((definition)) service is a fast and location specific version of the `definitions` relation in a ((Summary)).
+   * a ((reference)) service is a fast and location specific version of the `references` relation in a ((Summary)).
+   * an ((implementation)) service is a fast and location specific version of the `implementations` relation in a ((Summary)).
+* The ((documentSymbol)) service maps a source file to a pretty hierarchy for visualization in the "outline" view and "symbol search" features.
+* The ((lenses)) service discovers places to add "lenses" (little views embedded in the editor on a separate line) and connects commands to execute to each lense
+* The ((inlayHint)) service discovers plances to add "inlays" (little views embedded in the editor on the same line). Unlike ((lenses)) inlays do not offer command execution.
+* The ((execution)) service executes the commands registered by ((lenses)) and ((inlayHinter))s.
+* The ((actions)) service discovers palces to add "code actions" (little hints in the margin next to where the action is relevant) and connects ((CodeAction))s to execute when the users selects the action from a menu.
 
 Many language contributions received a ((Focus)) parameter. This helps to create functionality that
 is syntax-directed: relevant to the current syntactical constructs under the cursor.
 }
 data LanguageService
-    = parser(Parser parser)
-    | analyzer(Summarizer summarizer
+    = parsing(Tree (str _input, loc _origin) parser)
+    | analysis(Summary (loc _origin, Tree _input) summarizer
         , bool providesDocumentation = true
         , bool providesDefinitions = true
         , bool providesReferences = true
         , bool providesImplementations = true)
-    | builder(Summarizer summarizer
+    | build(Summary (loc _origin, Tree _input) summarizer
         , bool providesDocumentation = true
         , bool providesDefinitions = true
         , bool providesReferences = true
         , bool providesImplementations = true)
-    | outliner(Outliner outliner)
-    | lenses(OrderedLensDetector detector)
-    | inlayHinter(InlayHinter hinter)
-    | executor(CommandExecutor executor)
-    | documenter(FocusDocumenter documenter)
-    | definer(FocusDefiner definer)
-    | referrer(FocusReferrer reference)
-    | implementer(FocusImplementer implementer)
-    | actions(CodeActionContributor actions)
+    | documentSymbol(list[DocumentSymbol] (Tree _input) labeler)
+    | codeLense(lrel[loc src, Command lens] (Tree _input) detector)
```

Comment:
Nit: The corresponding request in LSP is `textDocument/codeLens` (without "e")

Reply:
Not a nit I think, it should be spelled `codeLens`. [Lense is not an English word](https://www.yourdictionary.com/articles/lense-vs-lens-difference).

Reply:
fixed! well spotted